### PR TITLE
Add connection string support + builder

### LINKED ISSUE #178: parse storage connection string
It would be great if there was a way to create a new storage account client from a storage account connection string. In .NET and other language SDKs, it is common to use the connection string.

It would simplify this a bunch by having two environment variables instead of four:
https://github.com/MindFlavor/AzureSDKForRust/blob/f30fb6e9ae540b067ca7ca39166093ff75690c8d/azure_sdk_storage_table/examples/copy_table.rs#L9-L23

### PR DESCRIPTION
**Pull request by guywaldman**
This PR:

- Adds a `ConnectionString` struct with most fields [taken from the .NET SDK](https://github.com/Azure/azure-storage-net/blob/ed89733dfb170707d65b7b8b0be36cb5bd6512e4/Lib/Common/CloudStorageAccount.cs#L63-L261) with examples
-  Adds a `from_connection_string` method on storage clients (Blob and Table) and updated the `copy_table` example
- Adds a `ConnectionStringBuilder` with examples

Closes #178

### CODE REVIEW

File: azure_sdk_storage_core/src/client.rs [hunk 1]
```diff
@@ -55,6 +47,59 @@ impl Client {
...
+                account: account.to_owned(),
+                key: key.to_owned(),
+                sas_token: None,
+                hc: client,
+                blob_uri: format!("https://{}.blob.core.windows.net", account),
+                table_uri: format!("https://{}.table.core.windows.net", account), 
+            }),
+            _ => {
+                return Err(AzureError::GenericErrorWithText(
+                    "Could not create an Azure Table client from the provided connection string. Please validate that you have specified the account name and means of authentication (key, SAS, etc.)."
```
Comment (guywaldman) [inline azure_sdk_storage_core/src/client.rs:96]:
```suggestion
                    "Could not create a storage client from the provided connection string. Please validate that you have specified the account name and means of authentication (key, SAS, etc.)."
```

File: azure_sdk_storage_core/src/connection_string.rs [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+            display("{}", msg)
+        }
+    }
+}
+
+impl From<ConnectionStringError> for AzureError {
+    fn from(err: ConnectionStringError) -> Self {
+        AzureError::GenericErrorWithText(err.to_string())
+    }
+}
```
Comment (guywaldman) [inline azure_sdk_storage_core/src/connection_string.rs:39]: @MindFlavor Haven't seen this done previously in the codebase, if there's a better way to convert between the errors let me know :)
Didn't want to create a whole new type in `AzureError`.

File: azure_sdk_storage_core/src/connection_string.rs [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+            EndpointProtocol::Https => "https",
+            EndpointProtocol::Http => "http",
+        };
+        write!(f, "{}", s)
+    }
+}
+
+/// A storage connection string.
+///
+/// The key are a subset of what is defined in the
```
Comment (guywaldman) [inline azure_sdk_storage_core/src/connection_string.rs:59]: Will change to:

```suggestion
/// The keys are a subset of what is defined in the
```

File: azure_sdk_storage_core/src/connection_string.rs [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+/// [.NET SDK](https://github.com/Azure/azure-storage-net/blob/ed89733dfb170707d65b7b8b0be36cb5bd6512e4/Lib/Common/CloudStorageAccount.cs#L63-L261).
+#[derive(Debug)]
+pub struct ConnectionString<'a> {
```
Comment (guywaldman) [inline azure_sdk_storage_core/src/connection_string.rs:62]: Missing some keys here, but I think these are the more popular ones.
Let me know if this is a blocker @MindFlavor.

As a side note, I considered doing a loose mapping here (basically making this a wrapper around a dictionary) and for example making `cs.account_name()` a method that returns the `AccountName` key from the mapping.
The pro is lazy-fetching of the properties, explicitness, and easier maintenance of these properties; the cons are an expensive `HashMap` and only getting errors upon key retrieval.

Could go either way here, not critical I think.

File: azure_sdk_storage_core/src/connection_string.rs (hunk 2 above)
Comment (MindFlavor) [reply azure_sdk_storage_core/src/connection_string.rs:39]: Sounds good! As of now the `AzureError` story needs some rethinking so I would not fret too much about it. Maybe we can encapsulate your error type in AzureError enum in the future (this probably means moving it to the `core ` crate). [https://github.com/MindFlavor/AzureSDKForRust/blob/69337ec4f6c62fbb87de158471c7b5e3b24b2eba/azure_sdk_core/src/errors.rs#L144-L285](https://github.com/MindFlavor/AzureSDKForRust/blob/69337ec4f6c62fbb87de158471c7b5e3b24b2eba/azure_sdk_core/src/errors.rs#L144-L285)

File: azure_sdk_storage_core/src/connection_string.rs (hunk 4 above)
Comment (MindFlavor) [reply azure_sdk_storage_core/src/connection_string.rs:62]: No it is not! It's great as it is now. We can always come back and add them in the future. As for the `HashMap` I prefer your approach: finding out an invalid connection string as soon as possible is important. Also, your code is elegant :+1: !

**Comment by MindFlavor** [comment]
Really, really impressive work! Thank you!

**Pull request merged**